plat/ti/k3low/common/drivers/pm/device/psc.c

```diff
@@ -0,0 +1,1431 @@
+		 (pd->use_count & TRACE_PM_VAL_MAX_PSC_DATA));
+
+	if ((pd->use_count++) != 0U) {
+		/* Nothing to do */
+	} else {
+		if ((psc->pd_data[idx].flags & PSC_PD_ALWAYSON) != 0U) {
+			/* Nothing to do */
+		} else {
+			/* Verify any previous transitions have completed */
+			psc_pd_wait(dev, pd);
+
+			if ((psc->pd_data[idx].flags & PSC_PD_DEPENDS) != 0U) {
+				psc_pd_get(dev, psc_idx2pd(psc,
+							   (pd_idx_t) psc->pd_data[idx].depends));
+			}
+
+			psc_pd_clk_get(&psc->pd_data[idx]);
+
+			pdctl = psc_read(dev, PSC_PDCTL(idx));
+
+			if ((pdctl & PDCTL_STATE_MASK) != PDCTL_STATE_ON) {
+				/* Avoid redundant power-up transitions */
+				pdctl &= ~PDCTL_STATE_MASK;
+				pdctl |= PDCTL_STATE_ON;
+
+				/* Note: This is a state machine reg */
+				psc_write(dev, pdctl, PSC_PDCTL(idx));
+
+				pd_initiate(dev, pd);
+				psc_pd_wait(dev, pd);
+			}
+
+			psc->data->pds_enabled |= (uint32_t) BIT(idx);
+		}
+	}
+}
+
+/**
+ * \brief Disable clocks necessary for power domain
+ *
+ * This function calls clk_put on each of the clocks listed
+ * under clock_dep for a given power domain. Some power domains
+ * require certain clocks to be running while the power domain
+ * is in transition or on.
+ *
+ * \param data
+ * The const data for the power domain
+ */
+static void psc_pd_clk_put(const struct psc_pd_data *data)
+{
+	uint32_t i;
+
+	for (i = 0U; i < ARRAY_SIZE(data->clock_dep); i++) {
+		struct clk *clkp = clk_lookup(data->clock_dep[i]);
+
+		if (clkp != NULL) {
+			clk_put(clkp);
+		}
+	}
+}
+
+void psc_pd_put(struct device *dev, struct psc_pd *pd)
+{
+	const struct psc_drv_data *psc = to_psc_drv_data(get_drv_data(dev));
+	uint32_t idx = psc_pd_idx(dev, pd);
+	uint32_t pdctl;
+
+	pm_trace(TRACE_PM_ACTION_PD_PUT,
+		 ((uint32_t) psc->psc_idx << TRACE_PM_VAL_PSC_SHIFT) |
+		 ((uint32_t) idx << TRACE_PM_VAL_PD_SHIFT) |
+		 (pd->use_count & TRACE_PM_VAL_MAX_PSC_DATA));
+
+	if ((--pd->use_count) != 0U) {
+		/* Nothing to do */
+	} else {
+		if ((psc->pd_data[idx].flags & PSC_PD_ALWAYSON) != 0U) {
+			/* Nothing to do */
+		} else {
+			/* Verify any previous transitions have completed */
+			psc_pd_wait(dev, pd);
+
+			pdctl = psc_read(dev, PSC_PDCTL(idx));
+			if ((pdctl & PDCTL_STATE_MASK) != PDCTL_STATE_OFF) {
+				/* Avoid redundant power-up transitions */
+				pdctl &= ~PDCTL_STATE_MASK;
+				pdctl |= PDCTL_STATE_OFF;
+				/* Note: This is a state machine reg */
+				psc_write(dev, pdctl, PSC_PDCTL(idx));
+
+				pd_initiate(dev, pd);
+				psc_pd_wait(dev, pd);
+
+			}
+
+			psc_pd_clk_put(&psc->pd_data[idx]);
+
+			if ((psc->pd_data[idx].flags & PSC_PD_DEPENDS) != 0U) {
+				psc_pd_put(dev, psc_idx2pd(psc,
+							   (pd_idx_t) psc->pd_data[idx].depends));
+			}
+
+			psc->data->pds_enabled &= ~((uint32_t) BIT(idx));
+		}
+	}
+}
+
+uint32_t psc_pd_get_state(struct device *dev, struct psc_pd *pd)
+{
+	uint32_t psc_ptstat = PSC_PTSTAT;
+	pd_idx_t psc_idx    = psc_pd_idx(dev, pd);
+
+	uint8_t state;
+	uint32_t ret;
```

Comment:
Idk why checkpatch did not complain on this one, but we can't have mixed declarations like these.